File: javascript/src/problems1_100/remove-duplicates-from-sorted-array.js
class Problem {
  removeDuplicates(nums) {
    throw new Error(`Not implemented (${nums}).`);
  }
}

class CreateCopy extends Problem {
  removeDuplicates(nums) {
    const distinct = new Set(nums);
    let i = 0;
    for (const num of distinct) {
      nums[i++] = num;
    }
    return distinct.size;
  }
}

class SinglePass extends Problem {
  removeDuplicates(nums) {
    let start = 1;
    let end = 1;
    while (end < nums.length) {
      if (nums[end] !== nums[end - 1]) {
        nums[start++] = nums[end];
      }
      end++;
    }
    return start;
  }
}

const SOLUTIONS = [new CreateCopy(), new SinglePass()];
export default SOLUTIONS;

File: python/src/problems1_100/remove_duplicates_from_sorted_array.py
from abc import ABC, abstractmethod
from typing import override


class Problem(ABC):
    @abstractmethod
    def remove_duplicates(self, nums: list[int]) -> int:
        pass


class CreateCopy(Problem):
    @override
    def remove_duplicates(self, nums: list[int]) -> int:
        distinct: set[int] = set(nums)
        i: int = 0
        for num in distinct:
            nums[i] = num
            i += 1
        return len(distinct)


class SinglePass(Problem):
    @override
    def remove_duplicates(self, nums: list[int]) -> int:
        start: int = 1
        end: int = 1
        while end < len(nums):
            if nums[end] != nums[end - 1]:
                nums[start] = nums[end]
                start += 1
            end += 1
        return start


SOLUTIONS: list[Problem] = [CreateCopy(), SinglePass()]
